// env-runner/dist/vite.mjs
//#region src/vite.ts
/**
* Create a Vite `HotChannel` from an env-runner's RPC hooks.
*
* Use on the **host side** to bridge env-runner IPC → Vite's DevEnvironment transport.
* Messages are namespaced by `envName` so multiple Vite environments can share one runner.
*/
function createViteHotChannel(hooks, envName) {
	const listeners = /* @__PURE__ */ new WeakMap();
	return {
		send: (data) => hooks.sendMessage({
			...data,
			viteEnv: envName
		}),
		on: (event, handler) => {
			if (event === "connection") return;
			const listener = (value) => {
				if (value?.type === "custom" && value.event === event && value.viteEnv === envName) handler(value.data, { send: (payload) => hooks.sendMessage({
					...payload,
					viteEnv: envName
				}) });
			};
			listeners.set(handler, listener);
			hooks.onMessage(listener);
		},
		off: (event, handler) => {
			if (event === "connection") return;
			const listener = listeners.get(handler);
			if (listener) {
				hooks.offMessage(listener);
				listeners.delete(handler);
			}
		}
	};
}
/**
* Create a Vite `ModuleRunner` transport from worker-side IPC primitives.
*
* Use on the **worker side** to bridge worker IPC → Vite's `ModuleRunner` transport.
* Filters messages by `envName` so multiple Vite environments can share one IPC channel.
*/
function createViteTransport(sendMessage, onMessage, envName) {
	return {
		connect({ onMessage: onRunnerMessage }) {
			onMessage((payload) => {
				if (payload?.type === "custom" && payload.viteEnv === envName) onRunnerMessage(payload);
			});
		},
		send(payload) {
			sendMessage?.({
				...payload,
				viteEnv: envName
			});
		}
	};
}
//#endregion
export { createViteHotChannel, createViteTransport };
